File: java/quickSortStep.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;

public class quickSortStep {
    // Data class defined within the same file
    static class Data implements Comparable<Data> {
        int number;
        String letters;

        public Data(int number, String letters) {
            this.number = number;
            this.letters = letters;
        }

        @Override
        public int compareTo(Data other) {
            return Integer.compare(this.number, other.number);
        }

        @Override
        public String toString() {
            return number + "/" + letters;
        }
        }

        // List to store sorting steps
        static ArrayList<String> sortingSteps = new ArrayList<>();

        public static void quickSort(Data[] arr, int low, int high, int startRow, int endRow) {
        if (low < high) {
            int pi = partition(arr, low, high, startRow, endRow);
            // Record step after partition
            sortingSteps.add("pi=" + pi + " " + arraySliceToString(arr, startRow, endRow));
            quickSort(arr, low, pi - 1, startRow, endRow);
            quickSort(arr, pi + 1, high, startRow, endRow);
        }
        }

        private static int partition(Data[] arr, int low, int high, int startRow, int endRow) {
        Data pivot = arr[high];
        int i = low - 1;
        for (int j = low; j < high; j++) {
            if (arr[j].compareTo(pivot) <= 0) {
            i++;
            Data temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
            }
        }
        Data temp = arr[i + 1];
        arr[i + 1] = arr[high];
        arr[high] = temp;
        return i + 1;
        }

        private static String arraySliceToString(Data[] arr, int start, int end) {
        StringBuilder sb = new StringBuilder();
        for (int i = start; i <= end; i++) {
            sb.append(arr[i].toString());
            if (i < end) sb.append(", ");
        }
        return "[" + sb.toString() + "]";
        }

        public static void main(String[] args) {
        java.util.Scanner scanner = new java.util.Scanner(System.in);
        System.out.print("Enter the CSV file path: ");
        String filePath = scanner.nextLine();
        ArrayList<Data> dataList = new ArrayList<>();
        try (BufferedReader br = new BufferedReader(new FileReader(filePath))) {
            String line;
            while ((line = br.readLine()) != null) {
            String[] parts = line.split(",", 2);
            if (parts.length == 2) {
            int number = Integer.parseInt(parts[0].trim());
            String letters = parts[1].trim();
            dataList.add(new Data(number, letters));
            }
            }
        } catch (IOException | NumberFormatException e) {
            System.out.println("Error reading file: " + e.getMessage());
            return;
        }
        Data[] arr = dataList.toArray(new Data[dataList.size()]);

        System.out.print("Enter start row: ");
        int startRowInput = scanner.nextInt();
        System.out.print("Enter end row: ");
        int endRowInput = scanner.nextInt();

        // Convert to 0-based index
        int startRow = startRowInput - 1;
        int endRow = endRowInput - 1;

        if (startRow < 0 || endRow >= arr.length || startRow > endRow) {
            System.out.println("Invalid start or end row.");
            return;
        }

        // Add the initial array slice before sorting steps
        sortingSteps.add(arraySliceToString(arr, startRow, endRow));

        long startTime = System.nanoTime();
        quickSort(arr, startRow, endRow, startRow, endRow);
        long endTime = System.nanoTime();
        double durationMs = (endTime - startTime) / 1_000_000.0;

        String outputFileName = String.format("quick_sort_step_%d_%d.txt", startRow + 1, endRow + 1);
        try (java.io.BufferedWriter bw = new java.io.BufferedWriter(new java.io.FileWriter(outputFileName))) {
            bw.write("Sorting steps for elements from row " + (startRow + 1) + " to " + (endRow + 1) + ":\n");
            for (String step : sortingSteps) {
            bw.write(step);
            bw.newLine();
            }
        } catch (IOException e) {
            System.out.println("Error writing file: " + e.getMessage());
        }
        }
    }

File: cpp/quick_sort_step.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <string>
#include <chrono>

struct Data {
    int number;
    std::string letters;

    bool operator<=(const Data& other) const {
        return number <= other.number;
    }
    std::string toString() const {
        return std::to_string(number) + "/" + letters;
    }
};

std::vector<std::string> sortingSteps;

std::string arraySliceToString(const std::vector<Data>& arr, int start, int end) {
    std::ostringstream oss;
    oss << "[";
    for (int i = start; i <= end; ++i) {
        oss << arr[i].toString();
        if (i < end) oss << ", ";
    }
    oss << "]";
    return oss.str();
}

int partition(std::vector<Data>& arr, int low, int high, int startRow, int endRow) {
    Data pivot = arr[high];
    int i = low - 1;
    for (int j = low; j < high; ++j) {
        if (arr[j] <= pivot) {
            ++i;
            std::swap(arr[i], arr[j]);
        }
    }
    std::swap(arr[i + 1], arr[high]);
    return i + 1;
}

void quickSort(std::vector<Data>& arr, int low, int high, int startRow, int endRow) {
    if (low < high) {
        int pi = partition(arr, low, high, startRow, endRow);
        sortingSteps.push_back("pi=" + std::to_string(pi) + " " + arraySliceToString(arr, startRow, endRow));
        quickSort(arr, low, pi - 1, startRow, endRow);
        quickSort(arr, pi + 1, high, startRow, endRow);
    }
}

int main() {
    std::string filePath;
    std::cout << "Enter the CSV file path: ";
    std::getline(std::cin, filePath);

    std::vector<Data> dataList;
    std::ifstream infile(filePath);
    if (!infile) {
        std::cout << "Error reading file: cannot open file." << std::endl;
        return 1;
    }
    std::string line;
    while (std::getline(infile, line)) {
        std::istringstream iss(line);
        std::string numStr, letters;
        if (std::getline(iss, numStr, ',') && std::getline(iss, letters)) {
            try {
                int number = std::stoi(numStr);
                dataList.push_back({number, letters});
            } catch (...) {
                // Ignore invalid lines
            }
        }
    }
    infile.close();

    int startRowInput, endRowInput;
    std::cout << "Enter start row: ";
    std::cin >> startRowInput;
    std::cout << "Enter end row: ";
    std::cin >> endRowInput;

    int startRow = startRowInput - 1;
    int endRow = endRowInput - 1;

    if (startRow < 0 || endRow >= (int)dataList.size() || startRow > endRow) {
        std::cout << "Invalid start or end row." << std::endl;
        return 1;
    }

    sortingSteps.push_back(arraySliceToString(dataList, startRow, endRow));

    auto startTime = std::chrono::high_resolution_clock::now();
    quickSort(dataList, startRow, endRow, startRow, endRow);
    auto endTime = std::chrono::high_resolution_clock::now();
    double durationMs = std::chrono::duration<double, std::milli>(endTime - startTime).count();

    std::string outputFileName = "quick_sort_step_" + std::to_string(startRow + 1) + "_" + std::to_string(endRow + 1) + ".txt";
    std::ofstream outfile(outputFileName);
    if (!outfile) {
        std::cout << "Error writing file: cannot open file." << std::endl;
        return 1;
    }
    outfile << "Sorting steps for elements from row " << (startRow + 1) << " to " << (endRow + 1) << ":\n";
    for (const auto& step : sortingSteps) {
        outfile << step << "\n";
    }
    outfile.close();

    return 0;
}
